[BB-2059] Remove deprecated city field from student features report

kaizoku opened the pull request:
This PR updates the student features report to pull from UserProfile.meta['city'] if it is set, in favor over `UserProfile.city`.

**Context**:
There are two distinct ways to store a city for users in edx-platform: one directly in `UserProfile.city`, and another in `UserProfile.meta`.

`UserProfile.meta` is used by adding `"city"` to the `extended_profile_fields`. This is then available via the account settings editor. The `UserProfile.city` field is not used or available via the account settings page.

Conversely to use a City field in the registration form, the `REGISTRATION_EXTRA_FIELDS["city"]` must be configured, and then entries are stored in `UserProfile.city`, which can't later be edited from the account settings page since it's [not included in the field set](https://github.com/edx/edx-platform/blob/0db4fb0ffe8ef568a9cae2b1c5da92065bbdd5e6/openedx/core/djangoapps/user_api/accounts/settings_views.py#L112).

Since the `UserProfile.meta['city']` field is not set unless specifically configured, we choose this over the `UserProfile.city` field when generating the student features report.

**Dependencies**: None

**Testing instructions**:

1. Add `"extended_profile_fields":"city"` (to the SiteConfiguration)[http://localhost:18000/admin/site_configuration/siteconfiguration/1/change/]
2. Set `{"city": "Test City"}` in the `UserProfile.meta` field for a (test user)[http://localhost:18000/admin/auth/user/5/change/].
3. Generate a report of student profiles with the "Download profile information as CSV" button in the instructor view for the [edX DemoX Course](http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/instructor#view-data_download).
4. Check that the "city" field is filled in from the value set in step #2.

**Reviewers**
- [ ] (@lgp171188)
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @kaizoku! I've created [OSPR-4114](https://openedx.atlassian.net/browse/OSPR-4114) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

lgp171188 commented:
@kaizoku, :+1: 
- [x] I tested that the `city` field is now stored in `UserProfile.meta`.
- [x] I read through the code
- [x] ~~I checked for accessibility issues~~ NA
- [x] ~~Includes documentation~~

natabene commented:
@kaizoku Thank you for your contribution. Please let me know once it is ready for our review.

giovannicimolin commented:
@natabene I've added this to the OSPR doc.

kaizoku commented:
@natabene thanks! This is ready for review.

natabene commented:
@giovannicimolin I am lining this up for our review.

kaizoku commented:
@natabene, is this still in the queue for review?

natabene commented:
@kaizoku Yes, it is still waiting for a review. We have very little bandwidth right now, so heads up it may take more time.

kaizoku commented:
@natabene, understood. Thanks for the update!

marcotuts commented:
I'm moving this back to product review - It isn't clear to me what the impact here is between city/country - Can you help me understand what the impact is to the reports with this change?

kaizoku commented:
@marcotuts, since the city field is no longer updated in the UserProfile serializer (https://github.com/edx/edx-platform/commit/ae0333cb61108386e16c7acbe3a9c6af483619e6) or from the account settings view, this just removes it from the student profile report.

The side effect of the revert is that country isn't added as the final field of the report. I could add another commit to move country to the final field similar to the what the commit being reverted in this PR does, but still leave out the city field.

marcotuts commented:
Sounds good to me. 

If the change currently results in the removal of the country field from the report I'd advocate for keeping that so we don't alter any workflow expectations here around that field, but removing the unused city field makes sense to me. 

Pushing this past product review again thanks for the detail @kaizoku

kaizoku commented:
Thanks for checking into this @marcotuts. I can confirm that this change keeps the country field, since it's still the field in use from the Account page and within the platform.

natabene commented:
@bradenmacdonald Please feel free to continue with this as a core committer.

bradenmacdonald commented:
Will do!

bradenmacdonald commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -1351,12 +1352,6 @@ def get_students_features(request, course_id, csv=False):  # pylint: disable=red
-    query_features.append('city')
-    query_features_names['city'] = _('City')
-    query_features.append('country')
-    query_features_names['country'] = _('Country')
```
@kaizoku I don't know if this really matters, but I think it would be better to leave the Country field at the end, just to keep the changes here as minimal as possible. In other words, instead of reverting that past commit, just remove the city field.

kaizoku replied:
@bradenmacdonald  Yep, that makes sense from a total change perspective and to maintain backwards compatibility as much as possible.
I added another commit which re-introduces the country field at the end instead. Should I merge these two to remove the revert so we just have one commit here?

bradenmacdonald replied:
Perfect, thanks! Yes, please squash to one commit and rebase, then I can approve and get this merged.

kaizoku replied:
Thanks @bradenmacdonald, I've squashed, written a better commit message, and rebased against master. This should be ready for merge now.

bradenmacdonald commented:
@kaizoku 

I tested this just now but couldn't confirm this:

> The UserProfile.city field is no longer used, in favor of adding the city to the UserProfile.meta JSON blob. 

In fact, it seems to still be used.

I set `REGISTRATION_EXTRA_FIELDS["city"]` to `"required"`, then went to http://localhost:18000/register and created a new account, including a City field during registration, and then found this:

```
# ./manage.py lms shell
>>> from django.contrib.auth import get_user_model
>>> User = get_user_model()
>>> User.objects.get(pk=10).profile.city
'Bobsville'
```

(the `meta` field is empty btw).

So, it seems like `UserProfile.city` _is_ still used? Or what am I missing?

kaizoku commented:
Thanks for catching that @bradenmacdonald. It's possible then that certain configurations can still use the `UserProfile.city` field then.

From the perspective of the account settings editor, the `meta.city` field is the only one which can be shown, which is why I thought this was deprecated.
Since [city isn't in the regular field set](https://github.com/edx/edx-platform/blob/0db4fb0ffe8ef568a9cae2b1c5da92065bbdd5e6/openedx/core/djangoapps/user_api/accounts/settings_views.py#L112), but it is part of the [extended field set](https://github.com/edx/edx-platform/blob/0db4fb0ffe8ef568a9cae2b1c5da92065bbdd5e6/openedx/core/djangoapps/user_api/accounts/settings_views.py#L236), a city field won't show up in the account settings page unless it's in the `extended_profile_fields`.

This may not be as clear cut as I thought then. Of course `UserProfile.city` can still be [included in the `student_profile_download_fields`](https://github.com/edx/edx-platform/blob/0db4fb0ffe8ef568a9cae2b1c5da92065bbdd5e6/lms/djangoapps/instructor/views/api.py#L1159), but it will change the default if we merge this.

I'm not sure what the intended field is to be used from the current code or comments, or this might just be a byproduct of separate developers adding things at different times which has led to two different places to store the field here. Should we attempt to merge these two somehow so we don't have two separate fields in use?

bradenmacdonald commented:
> I'm not sure what the intended field is to be used from the current code or comments, or this might just be a byproduct of separate developers adding things at different times which has led to two different places to store the field here. Should we attempt to merge these two somehow so we don't have two separate fields in use?

Hmm, maybe someone on @edx/teaching-and-learning could tell us, or tell us who to ask?

What we've found is that turning on `REGISTRATION_EXTRA_FIELDS["city"]` lets the user specify their city during account registration, and stores the city value in `UserProfile.city`. On the other hand, turning on `city` in `extended_profile_fields` lets users set their city from the account settings editor, but saves it into a different place, `UserProfile.meta`. Should we consolidate those?

bradenmacdonald commented:
@kdmccormick I tried to ping the TNL team in my last comment ^ but I'm not sure if team pings work properly like that. Would you know who I could ask about this? ("city" works differently on reg form vs. user profile)

kdmccormick commented:
@bradenmacdonald Yeah, the team pings work. Generally our on-call person is supposed to respond to them, so that's on me.

I'm clueless about this; I'll bring it up at standup and see if anyone else has an idea.

kdmccormick commented:
While I do some digging, @kaizoku do you mind updating the PR title and description? Seems like this isn't actually a revert, and the change is that we're removing the unused `UserProfile.city` field from instructor reports.

kaizoku commented:
Sure thing @kdmccormick, I wrote that when this was originally just a revert, but this PR has changed a bit since then.

I updated the description to match the current state of the PR.

kdmccormick commented:
It looks like `UserProfile.city` was originally added for use in a custom instance that we operate: https://github.com/edx/edx-platform/pull/2285. You may recognize the PR author :)

I'll ask internally whether that instance still needs the `UserProfile.city` field in the student report. While I do that, could you provide some context on why this change is being made?

Thanks all for your diligence so far on this.

kaizoku commented:
@kdmccormick haha I did notice when checking git blame that I know the author well. ;)

We're hoping to change this as one of our clients uses the city field in the registration form an account settings editor by adding "city" to `extended_profile_fields`, but this then caused some confusion since there are two separate entries for cities in the CSVs created by the student features report from the instructor dashboard.

I thought at the time it was an entirely deprecated field, so we should just remove it, but it's not quite so straightforward now that we realize there may be others still using the `UserProfile.city` field.

kdmccormick commented:
@kaizoku I applaud your effort to make things more consistent! Unfortunately this looks like one of those instances where two teams developed separate versions of the same feature, and both are still in use.

From the settings of our custom instance:
```
    REGISTRATION_EXTRA_FIELDS:
        ...
        city: 'required'
```

So, going forward:
* If you wanted to be ambitious and continue trying to remove the old field, you make a database migration that copies that value of `UserProfile.city` to `UserProfile.meta`, updates all references to `city`, and then another migration that drops the `city` column.
* If not, I'd support a PR that made the report first check `UserProfile.meta`, then falling back to `UserProfile.city` if it is missing/falsy. For good measure you could even leave a comment on `UserProfile.city` saying that it is deprecated in favor of of a field on `UserProfile.meta`.

kdmccormick commented:
Also, heads up that we're now in a holiday merge freeze until 1/4 for everything other than critical fixes.

bradenmacdonald commented:
> Unfortunately this looks like one of those instances where two teams developed separate implementations of the same feature, and both are still in use.

That's what I was afraid of, but also expecting. Thanks so much for looking into this for us in such detail, @kdmccormick !

kdmccormick commented:
No problem!

natabene commented:
@bradenmacdonald @kdmccormick Can you please advise on next steps for this PR?

kaizoku commented:
Thanks for checking in @natabene. I think we'll be taking approach #2 from @kdmccormick's [suggestions](https://github.com/edx/edx-platform/pull/23171#issuecomment-750338657). We're just clearing the budget to go ahead with implementing the approach.

kaizoku commented:
@natabene, no word yet. I just sent another ping about it though, perhaps the last one was missed.
I'll report back here when I hear about it.

kaizoku commented:
We got approval for the budget. I plan to rebase this and go ahead with implementing [the second of these suggestions](https://github.com/edx/edx-platform/pull/23171#issuecomment-750338657) in our next sprint (next week).

lgp171188 commented on lms/djangoapps/instructor_analytics/basic.py:
```diff
@@ -151,6 +152,9 @@ def extract_student(student, features):
             for meta_feature, meta_key in meta_features:
                 student_dict[meta_feature] = meta_dict.get(meta_key)
 
+            if include_city_column and 'city' in meta_dict:
```
@kaizoku, this doesn't fall back to `UserProfile.city` as @kdmccormick had recommended. If there isn't a reason for doing so, can you update this?

> If not, I'd support a PR that made the report first check UserProfile.meta, then falling back to UserProfile.city if it is missing/falsy. For good measure you could even leave a comment on UserProfile.city saying that it is deprecated in favor of of a field on UserProfile.meta.

kaizoku replied:
@lgp171188 , thanks for the catch. I tested with an empty UserProfile.meta but forgot to test with UserProfile.meta["city"] set to an empty string or the like.

This falls back to UserProfile.city if UserProfile.meta has no key for "city", but doesn't check if the key is falsy. I updated the PR to fall back to UserProfile.city if UserProfile.meta['city'] is an empty string or is completely unset.

lgp171188 commented on lms/djangoapps/instructor_analytics/basic.py:
```diff
@@ -151,6 +152,10 @@ def extract_student(student, features):
             for meta_feature, meta_key in meta_features:
                 student_dict[meta_feature] = meta_dict.get(meta_key)
 
+            meta_city = meta_dict.get('city')
+            if include_city_column and meta_city:
+                student_dict['city'] = meta_city
+
```
@kaizoku, with this change, wouldn't we have the same value in `student_dict['meta.city']` (populated in the `for` loop above) and `student_dict['city']`?

kaizoku replied:
@lgp171188 , only if we include `meta.city` and `city` in the query features. Only "city" is configured by default, if both are manually configured, this might be an issue, but should we be handling that scenario? As it stands, end users won't have to configure any options to get the city, it will always be there. So adding `meta.city` to the `student_profile_download_fields` would in a way be a misconfiguration.

natabene commented:
@kaizoku @lgp171188 Just checking what the status of this PR is. Is this ready for a final review by @bradenmacdonald ?

kaizoku commented:
Yes @natabene, we can go ahead with core committer review if @bradenmacdonald is up for it.
Guruprasad is still on vacation but I think I covered his last question here.

bradenmacdonald reviewed (changes requested):
@kaizoku  From the code alone, it's not obvious why this is being done this way, so can you please:

1. Add a comment to the code itself explaining the context here and/or linking to this PR? (e.g. `# There are two separate places where the city value can be stored, one used by account settings and the other used by the registration form. If the account settings value (meta.city) is set, it takes precedence.`
2. Rebase and squash to one commit with a conventional commit message.

Since @lgp171188 hasn't approved this fully and I'm not sure if he tested it, I will then do a manual test, approve, and merge.

arch-bom-gocd-alerts commented:
📣 💥 Heads-up: You must either **rebase onto master** or **merge master into your branch** to avoid breaking the build.

We recently removed diff-quality and introduced lint-amnesty. This means that the automated quality check that has run on your branch doesn't work the same way it will on master. If you have introduced any quality failures, they might pass on the PR but then break the build on master.

This branch has been detected to not have commit 2e335653 as an ancestor. Here's how to see for yourself:

```
git merge-base --is-ancestor 2e335653 josh/revert_city_student_features_report && echo "You're all set" || echo "Please rebase onto master or merge master to your branch"
```

If you have any questions, please reach out to the Architecture team (either #edx-shared-architecture on Open edX Slack or #architecture on edX internal).

kaizoku commented:
Thanks for the review and suggestions @bradenmacdonald. I used your suggested comment as it's pretty comprehensive for a concise comment. I rebased, squashed, and updated the commit message with some more details from our discussion here and a reference to this PR for all of the details.

I'm not positive if this qualifies as a "fix" or "feat" tag for the conventional commit type. I used "fix" since while it modifies a feature it seems more like a fix.

bradenmacdonald reviewed (approved):
:+1:

- [x] I tested this: set profile city, meta city, and both fields for three different users, confirmed that the report reads from both fields but prefers meta.
- [x] I read through the code
- [x] I checked for accessibility issues: n/a
- [x] Includes documentation: now has inline comment, yes

edx-webhook commented:
@kaizoku 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/23171&entry.1484655318=https://github.com/kaizoku&entry.752974735=2020-02-24+08:02&entry.1917517419=2021-06-16+18:26&entry.2133058324=Yes) so we can improve your experience in the future.
